File: code/threadsafe_queue/origin/threadsafe_queue_test.py
import threading
import time
from queue import Queue

def put_id():
    i = 0
    while True:
        i = i + 1
        print("添加数据", i, id_queue.qsize())
        time.sleep(0.1)
        id_queue.put(i)

def get_id(m):
    while True:
        i = id_queue.get()
        print("线程", m, '取值', i)


if __name__ == "__main__":


    id_queue = Queue(20)

    Th1 = threading.Thread(target=put_id, )

    Th2 = threading.Thread(target=get_id, args=(2, ))
    Th3 = threading.Thread(target=get_id, args=(3, ))
    Th5 = threading.Thread(target=get_id, args=(4, ))
    Th4 = threading.Thread(target=get_id, args=(5, ))

    Th2.start()
    Th1.start()

    Th3.start()
    Th4.start()
    Th5.start()

File: code/threadsafe_queue/recipe-03/threadsafe_queue_test.cpp
#include <thread>
#include <chrono>
#include <iostream>

#include "threadsafe_queue.hpp"

void put_id(threadsafe_queue<int>& queue) {
    int i = 0;
    while (true) {
        i = i + 1;
        std::cout << "添加数据 " << i << std::endl;
        std::this_thread::sleep_for(std::chrono::milliseconds(100));
        queue.push(i);
    }
}

void get_id(int thread_id, threadsafe_queue<int>& queue) {
    while (true) {
        int i;
        queue.wait_and_pop(i);
        std::cout << "线程: " << thread_id << " 取值 " << i << std::endl;
    }
}

int main() {
    threadsafe_queue<int> id_queue;
    auto Th1 = std::thread(put_id, std::ref(id_queue));

    auto Th2 = std::thread(get_id, 2, std::ref(id_queue));
    auto Th3 = std::thread(get_id, 3, std::ref(id_queue));
    auto Th5 = std::thread(get_id, 4, std::ref(id_queue));
    auto Th4 = std::thread(get_id, 5, std::ref(id_queue));

    Th2.join();
    Th1.join();

    Th3.join();
    Th4.join();
    Th5.join();
}
